File: prep/coreprep/string/rabin_karp.rs
use std::cmp::max;
use std::io;

fn get_num() -> i32 {
    let mut input = String::new();

    io::stdin()
        .read_line(&mut input)
        .expect("failed to read number");
    input.trim().parse::<_>().unwrap()
}

fn get_string() -> String {
    let mut input = String::new();

    io::stdin()
        .read_line(&mut input)
        .expect("failed to read string");
    input.trim().to_owned()
}

fn rabin_karp(s: &str, t: &str) -> Vec<usize> {
    let m = 53;
    let p = 1000000009;

    let slen = s.len();
    let tlen = t.len();

    let mut ppow = vec![0; max(slen, tlen)];

    ppow[0] = 1;
    for i in 1..ppow.len() {
        ppow[i] = (ppow[i - 1] * p) % m;
    }

    let mut shash = 0;
    for (i, b) in s.bytes().enumerate() {
        shash = (shash + (b - b' ' + 1) as i64 * ppow[i]) % m;
    }

    let mut thashes = vec![0; tlen + 1];
    for (i, b) in t.bytes().enumerate() {
        thashes[i + 1] = (thashes[i] + (b - b' ' + 1) as i64 * ppow[i]) % m;
    }

    let mut occurrences = Vec::new();
    for i in 0..tlen - slen + 1 {
        let curr_hash = (thashes[i + slen] + m - thashes[i]) % m;

        if curr_hash == shash * ppow[i] % m {
            occurrences.push(i);
        }
    }

    occurrences
}

fn main() {
    let n = get_num();

    for _ in 0..n {
        let s = get_string();
        let t = get_string();

        let res = rabin_karp(&s, &t);
        println!("s = {s}, t = {t}");

        if res.is_empty() {
            println!("-1");
        } else {
            println!("{}", res.len());
            for pos in res {
                print!("{} ", pos);
            }
            println!();
        }
    }
}

File: prep/coreprep/string/RabinKarp.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class RabinKarp {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      in.nextLine();

      while (n-- > 0) {
        String s = in.nextLine().trim();
        String t = in.nextLine().trim();

        var res = rabinKarp(s, t);
        if (res.isEmpty()) {
          System.out.println(-1);
        } else {
          System.out.println(res.size());
          for (var pos : res) {
            System.out.printf("%d ", pos);
          }
          System.out.println();
        }
      }
    }
  }

  private static List<Integer> rabinKarp(String s, String t) {
    int m = 53;
    int p = (int)1e9 + 9;
    int slen = s.length(), tlen = t.length();
    long[] ppow = new long[Math.max(slen, tlen)];

    ppow[0] = 1;
    for (int i = 1; i < ppow.length; i++) {
      ppow[i] = (ppow[i - 1] * p) % m;
    }

    long shash = 0;
    for (int i = 0; i < slen; i++) {
      shash = (shash + (s.charAt(i) - ' ' + 1) * ppow[i]) % m;
    }

    if (shash < 0) {
      shash += m;
    }

    long[] thashes = new long[tlen + 1];
    for (int i = 0; i < tlen; i++) {
      thashes[i + 1] = (thashes[i] + (t.charAt(i) - ' ' + 1) * ppow[i]) % m;
    }

    for (int i = 0; i < thashes.length; i++) {
      if (thashes[i] < 0) {
        thashes[i] += m;
      }
    }

    List<Integer> occurrences = new ArrayList<>();
    for (int i = 0; i < tlen - slen + 1; i++) {
      long currHash = (thashes[i + slen] + m - thashes[i]) % m;

      if (currHash == shash * ppow[i] % m) {
        occurrences.add(i);
      }
    }

    return occurrences;
  }
}